src/nupic/support/group_by.py

```diff
@@ -0,0 +1,161 @@
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+
+
+class GroupByGenerator(object):
+  """ A custom generator used by groupby to allow for len() calls on the
+      generator returned.
+  """
+
+  __slots__ = ['iterable', 'beginning', 'length', 'end', 'index']
+
+  def __init__(self, iterable, beginning, length):
+    self.iterable = iterable
+    self.beginning = beginning
+    self.length = length
+    self.end = self.beginning + self.length
+    self.index = 0
+
+
+  def __iter__(self):
+    for i in xrange(self.beginning, self.end):
+      self.index = i
+      yield self.iterable[i]
+
+
+  def __len__(self):
+    return self.length - self.index
+
+
+def groupby(lis, fn):
+  """ A custom implementation of itertools.groupby that doesn't reuse the
+      generator objects, allowing for groupByN to not have to capture the
+      group in a list.
+
+  @param lis (list) the list to perform the groupby on
+  @param fn  (function) the key function to perform the groupings
+
+  @return (tuple) A (key, GroupByGenerator) tuple where the first value is the
+                  key of the group, and second value is the generator that
+                  generates the values in the group.
+  """
+  length = 1
+  beginning = 0
+  l = len(lis)
+
+  if l > 0:
+    key = fn(lis[0])
+    for i in xrange(1, len(lis)):
+      val = fn(lis[i])
+
+      if val != key:
+        yield (key, GroupByGenerator(lis, beginning, length))
+        length = 1
+        beginning = i
+        key = val
+      else:
+        length += 1
+
+  if l > 0: # yield last group
+    yield (key, GroupByGenerator(lis, beginning, l - beginning))
+
+
+def groupByN(*args):
+  """ A utility function wrapper based on groupby in itertools. Allows to
+      walk across n sorted lists with respect to their key functions
+      and yields a tuple of n lists of the members of the next *smallest*
+      group.
+
+  @param args (list) a list of arguments alternating between sorted lists and
+                       their respective key functions. The lists should be
+                       sorted with respect to their key function.
+
+  @return (tuple) a n + 1 dimensional tuple, where the first element is the
+                    key of the group and the other n entries are lists of
+                    objects that are a member of the current group that is being
+                    iterated over in the nth list passed in. Note that this
+                    is a generator and a n+1 dimensional tuple is yielded for
+                    every group.
+
+  Notes: Read up on groupby here:
+         https://docs.python.org/dev/library/itertools.html#itertools.groupby
+
+         itertools.groupby was not used and the justification can be seen
+         here https://github.com/numenta/nupic/pull/3254#discussion_r74164564
```

Comment:
The docstring for `groupby` should answer this question. We probably shouldn't link to a github discussion. Better to describe it succinctly here.

Fundamentally, I think `itertools.groupby` is more constrained because it has to support arbitrary iterables. You couldn't pass `(i for i in xrange(10))` into our custom `groupby`, but you can pass it into `itertools.groupby`.

And I think now that we're not evaluating the whole groupby result immediately in `groupByN`, we could use the built-in `itertools.groupby`, though I'm not necessarily suggesting this. The benefits we're getting are:

- Support for `len`
- Support for arbitrary evaluation of the nested generators returned by `groupByN` (though we're not currently using this)